File: Data Structures & Algorithms/balanced-binary-tree/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {boolean}
     */
    isBalanced(root) {
        let res = [true]
        this.traverse(root,res)
        return res[0]

    }

    traverse(root,res){
        console.log("ATTRAVERSO: ",root)
        if(!root) return 0

        let dl = this.traverse(root.left,res)
        let dr = this.traverse(root.right,res)

        if(Math.abs(dl-dr) > 1) res[0] = false

        return 1 + (dl > dr ? dl : dr)
    }
}

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public int GoodNodes(TreeNode root) {

        if(root is null) return 0;

        var res = 0;

        void traverse(TreeNode node, int max){

            if(node is null) return;

            if(node.val >= max){
                res++;
                max = node.val;
            }

            traverse(node.left, max);
            traverse(node.right, max);
        }

        traverse(root, -101);
        return res;
        
    }
}
